haskell.compiler.ghc{9102,9122}: drop

wolfgangwalther opened the pull request:
Drops all GHC minor versions that are currently:
- Not the latest minor version of the respective major version, and
- Not the latest version in a stackage major, and
- Not used for bootstrapping others (GHC 9.6.3 is used to bootstrap GHC 9.6.7, so can't drop that)

This applies our [Deprecation Policy](https://nixos.org/manual/nixpkgs/unstable/#ghc-deprecation-policy). Removal of GHC minor version has so far not appeared in Release Notes, so not adding these here either.

9.4, 9.6 and 9.8 drops already merged.

## Things done

- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

emilazy commented:
Fair enough. Although the arguments for keeping it seem sufficiently weak that “saving having to do another build to test whether it is actually blocking something” is enough for me, personally 😅

(This can target `master` too?)

wolfgangwalther committed:
haskell.compiler.ghc9102: drop

Latest 9.10.x minor release is 9.10.3, which is also in Stackage 24.12+.

Thus, dropping according to the GHC Deprecation Policy.

emilazy commented:
Probably one of:

* https://gitlab.haskell.org/ghc/ghc/-/commit/23789c2839fe403c085161b1dbbd762932750881
* https://gitlab.haskell.org/ghc/ghc/-/commit/0b4316fdc37cfd4f01e989e306991b3a5bf47002
* https://gitlab.haskell.org/ghc/ghc/-/commit/4cc84ccffb2928af3a9bf3ea6d0262684ac454d4

emilazy commented:
Perhaps we should just split this into two PRs, since dropping 9.12.1 seems uncontroversial? It looks good to me. But of course if you want to push the commit directly that seems fine as well.

wolfgangwalther committed:
ghc/common-hadrian: skip .1 on versionOlder checks

This makes it a tad clearer that this matches only the prior major
versions.

wolfgangwalther committed:
haskell.compiler.ghc9122: drop

Latest 9.12.x minor release is 9.12.4, which is also in Stackage Nightly
already, so will never be the latest in a Stackage major version.

Thus, dropping according to the GHC Deprecation Policy.

wolfgangwalther commented:
> It seems to me that we should review before branch off which compilers from 9.10/9.12 we can drop.

Next branch off coming up in a month or so - therefore picking up the discussion again. Kept the drop for 9.10.2 and added the drop for GHC 9.12.2 as well. These are the minor version drops we should do right now according to the GHC Deprecation Policy.

The only major version that could be dropped according to the policy would still be 9.4. The 9.6+ TH problem is solved now, but I guess we'll still want to keep 9.4 for cross-compiling reasons.

sternenseemann commented:
> The only major version that could be dropped according to the policy would still be 9.4. The 9.6+ TH problem is solved now, but I guess we'll still want to keep 9.4 for cross-compiling reasons.

Yes. As things stand I think 9.6 should be the next major version we remove (wasn't a very good version anyways), though our old friend Elm depends on that at the moment unfortunately.

I am not aware of any problems with 9.10.3 since we've discussed it last. https://github.com/NixOS/nixpkgs/pull/439323#issuecomment-3403658733 is probably not reason enough to keep it around on the stable branch.

9.12 is really in a sad state, but 9.12.3 is seems very usable since we've patched the miscompilation. 9.12.4 can't really compile profiling objects without crashing which is very unfortunate for us. It's apparently not a given that there will be a 9.12.5 and a fix for the profiling crashes is still pending for master last I checked. I'm unsure whether backporting that when it materializes or reverting stuff from the 9.12 branch will be the best course of action.

guibou commented:
> 9.12 is really in a sad state, but 9.12.3 is seems very usable since we've patched the miscompilation. 9.12.4 can't really compile profiling objects without crashing which is very unfortunate for us. It's apparently not a given that there will be a 9.12.5 and a fix for the profiling crashes is still pending for master last I checked. I'm unsure whether backporting that when it materializes or reverting stuff from the 9.12 branch will be the best course of action.

Note that 9.12.4 (and 9.14) also contains a bug in the RTS which could randomly crash / deadlock, see https://gitlab.haskell.org/ghc/ghc/-/merge_requests/15640. So better to keep 9.12.2 or 9.12.3, but avoid 9.12.4 for now.

wolfgangwalther commented:
> So better to keep 9.12.2 or 9.12.3, but avoid 9.12.4 for now.

This PR keeps 9.12.3, which is the default for 9.12 for us right now anyway. It only drops 9.12.2.

wolfgangwalther commented:
> As things stand I think 9.6 should be the next major version we remove (wasn't a very good version anyways), though our old friend Elm depends on that at the moment unfortunately.

Right. According to our policy, we can only remove 9.6 once we updated the default version to 9.12 anyway, so certainly not for 26.05.

I understand everything else you wrote essentially as an approval of this PR.
